sdk/python/kfp/dsl/_container_op.py

```diff
@@ -1240,6 +1241,17 @@ def _decorated(*args, **kwargs):
 
     self.pvolumes = {}
     self.add_pvolumes(pvolumes)
+    
+    self._cpu_request = None
+    self._memory_request = None
+
+  @property
+  def cpu_request(self):
```

Comment:
We already have 
[ContainerOp.container.set_cpu_request](https://github.com/kubeflow/pipelines/blob/f20d20c41d434a34e63de6b02b83846c286e3db5/sdk/python/kfp/dsl/_container_op.py#L334). (`ContainerOp.set_cpu_request`) works, but is deprecated.

See the [sample](https://github.com/kubeflow/pipelines/blob/cc83e1089b573256e781ed2e4ac90f604129e769/sdk/python/tests/dsl/container_op_tests.py#L77).

You just need to modify how they compile rather than adding new properties.

For example: After ContainerOp has been compiled to a container template, you can take the `resources` section and move it to `podSpecPatch` (unconditionally or only if it contains PipelineParams)

Reply:
Note that this is a read-only property (there's no setter), just to prevent changing the value without calling the `set_cpu_request`/`add_cpu_request`.

But, +1 if we have `set_cpu_request` already, we should just reuse that interface and update its implementation.

@Ark-kun, I wasn't aware of this history, but why are they deprecated?

Reply:
>@Ark-kun, I wasn't aware of this history, but why are they deprecated?

Sorry for confusion. I closed the parentheses too early.
`ContainerOp.<some container method>` was deprecated in favor of `ContainerOp.container.<some container method>`. This was done for maintainability. `ContainerOp.*` is maintained manually while `ContainerOp.container.*` is just Kubernetes' `V1Container`, so it supports all Kubernetes container features automatically. Also, `ContainerOp` can have multiple containers (init containers and sidecars), so there is a benefit to be able to set the properties of all those containers in the same way.

Reply:
>Since this change actually uses the podSpecPatch in argo maybe we should allow the user to supply the patch them self and only add it as the podsSpecPatch.

`podsSpecPatch` is an Argo-specific feature. I'm not sure we should directly expose it.

Reply:
>since this should wrap over the container object I am not sure we want to combine the implementation with the runtime requirements which will rely on podSpecPatch.

Let's look at this from the architecture layering perspective:
* The DSL gives full access to K8s container properties (including limits and requests). I think that's what we expose to the user and what we should continue exposing.
* The DSL should only expose Kubernetes or KFP features, not direct Argo features.
* The DSL allows using placeholders (`PipelineParam`) instead of static values.
* From the perspective of the DSL, there is nothing wrong in using placeholders in limits/requests.
* The current compiler translates the placeholders in a way that violates the V1Container schema used by Kubernetes and Argo. The issue exists on the compiler level, not the DSL level.
* The way the compiler translates the placeholders should be fixed.
* The compiler can be fixed by detecting cases where directly putting the string placeholders in the `V1Container` attributes would result in invalid Workflow and put the placeholders in Argo's `podSpecPatch` instead.

Reply:
Thanks for the clarification, I will update the PR.